src/main/java/input_feed/distributed/TextToDistributedR1CS.java

```diff
@@ -0,0 +1,220 @@
+        int numAuxiliary = Integer.parseInt(constraintParameters[1]);
+        int numConstraints = Integer.parseInt(constraintParameters[2]);
+
+        // Need at least one constraint per partition!
+        assert(numConstraints >= this.config().numPartitions());
+
+        final ArrayList<Integer> partitions = constructPartitionArray(this.config().numPartitions(), numConstraints);
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationA = distributedCombination(
+                this.filePath() + ".a", partitions, numConstraints, false);
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationB = distributedCombination(
+                this.filePath() + ".b", partitions, numConstraints, false);
+        JavaPairRDD<Long, LinearTerm<FieldT>> linearCombinationC = distributedCombination(
+                this.filePath() + ".c", partitions, numConstraints, this.negate());
+
+        linearCombinationA.count();
+        linearCombinationB.count();
+        linearCombinationC.count();
+
+        final R1CSConstraintsRDD<FieldT> constraints = new R1CSConstraintsRDD<>(
+                linearCombinationA,
+                linearCombinationB,
+                linearCombinationC,
+                numConstraints);
+
+        return new R1CSRelationRDD<>(constraints, numInputs, numAuxiliary);
+    }
+
+    @Override
+    public Tuple2<Assignment<FieldT>, JavaPairRDD<Long, FieldT>> loadWitness() {
+
+        final Assignment<FieldT> serialAssignment = new Assignment<>();
+        int numInputs = -1;
+        int numAuxiliary = -1;
+
+        try{
+            String[] constraintParameters = new BufferedReader(
+                    new FileReader(this.filePath() + ".problem_size")).readLine().split(" ");
+
+            numInputs = Integer.parseInt(constraintParameters[0]);
+            numAuxiliary = Integer.parseInt(constraintParameters[1]);
+
+            BufferedReader br = new BufferedReader(
+                    new FileReader(this.filePath() + ".witness"));
+
+            String[] splitWitness = br.readLine().split("\\s+");
+            int count = 0;
+            for (String next: splitWitness) {
+                final FieldT value = this.fieldParameters().construct(next);
+                serialAssignment.add(value);
+                count++;
+            }
+
+            assert (count == numInputs + numAuxiliary);
+            br.close();
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+
+        int totalSize = numInputs + numAuxiliary;
+
+        final int numExecutors = this.config().numExecutors();
+        final ArrayList<Integer> assignmentPartitions = new ArrayList<>();
+        for (int i = 0; i < numExecutors; i++) {
+            assignmentPartitions.add(i);
+        }
+        if (totalSize % 2 != 0) {
+            assignmentPartitions.add(numExecutors);
+        }
+
+        JavaPairRDD<Long, FieldT> distributedAssignment = this.config().sparkContext()
```

Comment:
Do we have to parallelize this distribution? It seems like we are shipping a large list of variables to the nodes, and they ship us back their fraction of that list. There is not actually any logic going on.

Would distributed proof generation work with serial assignment loading? This is create a JavaPairRDD from a local object.

Reply:
I am not sure I understand the question here. distributed proof expects JavaPairRDD as assignment, while serial proof expects some sort of arrayList

Reply:
Yes, but could you create a JavaPairRDD just on the master, without distributing anything (given that there is no logic happening in each worker currently)? It feels like we are just shipping data over the network for no good reason.